File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        numSum = sum(nums)
        if numSum % 2 != 0:
            return False
        resultSum = numSum // 2
        checkMap = {}
        def recursion(nums, resultSum) -> bool:
            if resultSum in checkMap:
                return checkMap[resultSum]
            while len(nums) > 0:
                num = nums.pop(0)
                if resultSum - num == 0:
                    return True
                if resultSum - num < 0:    
                    continue
                if recursion(list(nums), resultSum - num):
                    return True
            checkMap[resultSum] = False
            return False
        
        return recursion(nums, resultSum)

File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.dart
class Solution {
  bool canPartition(List<int> nums) {
    final sum = nums.reduce((a, b) => a + b);
    if (sum % 2 == 1) return false;
    final half = sum ~/ 2;
    var dp = List.filled(half + 1, false);
    dp[0] = true;
    for (int i = 0; i < nums.length; i++){
      final temp = List.filled(half + 1, false);
      temp[0] = true;
      for (int j = 0; j <= half; j++){
        if (!dp[j]) continue;
        temp[j] = true;
        if (j + nums[i] > half) break;
        temp[j + nums[i]] = true;
      }
      dp = temp;
    }
    return dp[half];
  }
}